data-prepper-core/src/test/java/org/opensearch/dataprepper/pipeline/router/RouterCopyRecordStrategyTests.java

```diff
@@ -18,42 +18,99 @@
 import java.util.Collection;
 import java.util.Collections;
 import java.util.Map;
+import java.util.Iterator;
 import java.util.HashMap;
+import java.util.List;
+import java.util.ArrayList;
 import java.util.Set;
 import java.util.UUID;
 import java.util.stream.Collectors;
 import java.util.stream.IntStream;
 
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.when;
+import static org.mockito.Mockito.lenient;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.ArgumentMatchers.any;
 import static org.hamcrest.Matchers.equalTo;
 import static org.hamcrest.CoreMatchers.sameInstance;
 import static org.hamcrest.CoreMatchers.not;
 import static org.hamcrest.MatcherAssert.assertThat;
+import static org.junit.jupiter.api.Assertions.assertTrue;
 import static org.junit.jupiter.api.Assertions.assertThrows;
 import static org.junit.Assert.assertFalse;
 import org.opensearch.dataprepper.model.event.Event;
 import org.opensearch.dataprepper.model.event.JacksonEvent;
+import org.opensearch.dataprepper.model.event.EventFactory;
+import org.opensearch.dataprepper.model.event.EventHandle;
+import org.opensearch.dataprepper.model.event.EventBuilder;
+import org.opensearch.dataprepper.model.event.EventMetadata;
+import org.opensearch.dataprepper.event.DefaultEventHandle;
+import org.opensearch.dataprepper.model.acknowledgements.AcknowledgementSetManager;
+import org.opensearch.dataprepper.model.acknowledgements.AcknowledgementSet;
 
 @ExtendWith(MockitoExtension.class)
 public class RouterCopyRecordStrategyTests {
 
     @Mock
     private DataFlowComponent<PipelineConnector> pipelineDataFlowComponent;
     private Collection<Record> recordsIn;
+    private Collection<Record> mockRecordsIn;
+    @Mock
+    private AcknowledgementSetManager acknowledgementSetManager;
+    @Mock
+    private AcknowledgementSet acknowledgementSet1;
+    @Mock
+    private EventFactory eventFactory;
+    @Mock
+    private EventBuilder eventBuilder;
+
+    private JacksonEvent event;
+
+    private Map<EventHandle, Integer> handleRefCount;
 
     private static class TestComponent {
     }
 
     @BeforeEach
     void setUp() {
-        recordsIn = IntStream.range(0, 10)
+        handleRefCount = new HashMap<>();
+        eventFactory = mock(EventFactory.class);
+        acknowledgementSetManager = mock(AcknowledgementSetManager.class);
```

Comment:
It would be ideal to have some tests which explicitly set this value to `null`. Otherwise, you may miss a `null` check on it in the main code and have a crash.

Reply:
An alternative is to create an `InactiveAcknowledgementSetManager` when it is not needed.

```
class InactiveAcknowledgementSetManager implements AcknowledgementSetManager {
  public static InactiveAcknowledgementSetManager getInstance() {
    // standard singleton work
  }

  public void acquireEventReference(final Event event) {
    // no-op
   }

  public AcknowledgementSet create(final Consumer<Boolean> callback, final Duration timeout) {
    throw new UnsupportedOperationException(...);
  }
}
```

Then you have in `Pipeline`:

```
(source.areAcknowledgementsEnabled()) ? acknowledgementSetManager : InactiveAcknowledgementSetManager.getInstance(), sinks);
```


The `InactiveAcknowledgementSetManager` class would be quite easy to unit test.

Reply:
Another option I thought of is to include a boolean in EventHandle that indicates if it should be acknowledged or not.

When the source first writes to the buffer, if there is an `AcknowledgementSet`, then the `EventHandle` could be set to have tracking. Otherwise, false.

Then you can always use `AcknowledgementSetManager`. It would look at the `EventHandle` and do the no-op itself if it should not be acknowledged.

Reply:
Let's go ahead with what you have here. Please fix this later.

Reply:
Will implement `InactiveAcknowledgementSetManager` in a different PR